### PR TITLE
WIP: switch to fully decomposed tokens internally

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

File: crates/ra_parser/src/grammar/expressions.rs [hunk 1]
```diff
@@ -212,52 +212,45 @@ struct Restrictions {
...
-            (T![>], T![=]) => return (5, Op::Composite(T![>=], 2)),
-            (T![<], T![<]) => return (9, Op::Composite(T![<<], 2)),
-            (T![>], T![>]) => return (9, Op::Composite(T![>>], 2)),
-            _ => (),
-        }
+#[rustfmt::skip]
+fn current_op(p: &Parser) -> (u8, SyntaxKind) {
+    const NOT_AN_OP: (u8, SyntaxKind) = (0, T![@]);
+    match p.current() {
+        T![|] if p.at(T![||])  => (3,  T![||]),
```
Comment (bjorn3) [inline crates/ra_parser/src/grammar/expressions.rs]: What do these numbers mean?

Comment (matklad) [reply crates/ra_parser/src/grammar/expressions.rs]: priority of a binary operator for a [pratt parser](https://www.oilshell.org/blog/2016/11/03.html). This really could use a comment though, thanks!

Comment (matklad) [reply crates/ra_parser/src/grammar/expressions.rs]: And, in general, I should say I really enjoy your drive-by comments on various commits, thanks! :)

**Comment by matklad** [comment]
bors try

**Comment by matklad** [comment]
I am going to merge this, although it regresses one macro-expansion tests, as this seems to be an overall better approach

bors

**Pull request merged**